README
Gedit Scroll Paste Fix
===================================

This plugin fixes GTK's scrolling behaviour when pasting text with the mouse.
By default when text is pasted in a text view using the mouse the view scrolls
to the current cursor position and not to where the text has been pasted. This
is extremely annoying specially when pasting text in another location that's far
away from the cursor.

Project homepage: http://github.com/potyl/gedit-scrollpaste

Building
========

This project uses autoconf & friends as its build toolchain. Building it should
be trivial once the GNU autoconf toolchain is installed.

In Debian based systems the build dependencies can be easily installed by
doing:

  sudo apt-get install build-essential autoconf libtool gnome-common
  sudo apt-get install libgtksourceview2.0-dev gedit-dev

To install on Fedora, make sure you have a working toolchain installed and the the build dependencies:

  sudo yum install gcc make autoconf automake libtool gettext
  sudo yum install gtksourceview-devel gedit-devel

For OS X the plugin can be either compiled for the macports versions or for
the native application installed through a native dmg image. For both versions
you first need to download and install gedit through the macports:

  sudo port install gedit gedit-plugins

If you have forked the code from github you will need to run the autoconf
programs in order to prepare build toolchain:

  ./autogen.sh

Once the toolchain scripts are created simply build the program with:

  ./configure --prefix=/usr && make && sudo make install

A faster way is to pass the arguments of ./configure directly to ./autogen.sh
this way the command becomes:

  ./autogen.sh --prefix=/usr && make && sudo make install

For installing the native OS X version of the plugin pass --enable-osx to the
configuration script:

  ./autogen.sh --enable-osx && make && sudo make install


ChangeLog
0.0.3 Thu Mar 24 22:43:48 CET 2011
	- Allow the plugin to be compiled for the native version of gedit on OS X.


0.0.2 Mon Apr 12 20:26:25 CEST 2010
	- Include the required files for autoconf.


0.0.1 Sat Nov 14 22:28:53 CET 2009
	- Initial release.


src/scrollpaste-plugin.h
/*
 * scrollpaste-plugin.h - Fix scrolling behavior after a paste
 *
 * Copyright (C) 2009 - Emmanuel Rodriguez
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifndef __SCROLLPASTE_PLUGIN_H__
#define __SCROLLPASTE_PLUGIN_H__

#include <glib.h>
#include <glib-object.h>
#include <gedit/gedit-plugin.h>

G_BEGIN_DECLS

/*
 * Type checking and casting macros
 */
#define TYPE_SCROLLPASTE_PLUGIN		(scrollpaste_plugin_get_type ())
#define SCROLLPASTE_PLUGIN(o)		(G_TYPE_CHECK_INSTANCE_CAST ((o), TYPE_SCROLLPASTE_PLUGIN, ScrollpastePlugin))
#define SCROLLPASTE_PLUGIN_CLASS(k)	(G_TYPE_CHECK_CLASS_CAST((k), TYPE_SCROLLPASTE_PLUGIN, ScrollpastePluginClass))
#define IS_SCROLLPASTE_PLUGIN(o)	(G_TYPE_CHECK_INSTANCE_TYPE ((o), TYPE_SCROLLPASTE_PLUGIN))
#define IS_SCROLLPASTE_PLUGIN_CLASS(k)	(G_TYPE_CHECK_CLASS_TYPE ((k), TYPE_SCROLLPASTE_PLUGIN))
#define SCROLLPASTE_GET_CLASS(o)	(G_TYPE_INSTANCE_GET_CLASS ((o), TYPE_SCROLLPASTE_PLUGIN, ScrollpastePluginClass))

/* Private structure type */
typedef struct _ScrollpastePluginPrivate	ScrollpastePluginPrivate;

/*
 * Main object structure
 */
typedef struct _ScrollpastePlugin		ScrollpastePlugin;

struct _ScrollpastePlugin
{
	GeditPlugin parent_instance;

	/*< private >*/
	ScrollpastePluginPrivate *priv;
};

/*
 * Class definition
 */
typedef struct _ScrollpastePluginClass	ScrollpastePluginClass;

struct _ScrollpastePluginClass
{
	GeditPluginClass parent_class;
};

/*
 * Public methods
 */
GType	scrollpaste_plugin_get_type	(void) G_GNUC_CONST;

/* All the plugins must implement this function */
G_MODULE_EXPORT GType register_gedit_plugin (GTypeModule *module);

G_END_DECLS

#endif /* __SCROLLPASTE_PLUGIN_H__ */



src/scrollpaste-plugin.c
/*
 * scrollpaste-plugin.c - Fix scrolling behavior after a paste
 *
 * Copyright (C) 2009 - Emmanuel Rodriguez
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include "scrollpaste-plugin.h"

#include <glib/gi18n-lib.h>
#include <gedit/gedit-debug.h>
#include <glib.h>

#define WINDOW_DATA_KEY	"ScrollpastePluginWindowData"

#define SCROLLPASTE_PLUGIN_GET_PRIVATE(object)	(G_TYPE_INSTANCE_GET_PRIVATE ((object), TYPE_SCROLLPASTE_PLUGIN, ScrollpastePluginPrivate))

struct _ScrollpastePluginPrivate
{
	/* Empty */
	gboolean dummy;
};

typedef struct _PasteData {
	GtkTextMark *paste_mark;
	GtkTextView *view;
} PasteData;


static void paste_data_free (gpointer data);

static void disconnect_from_view (GeditView *view, gpointer data);

static void connect_to_view (GeditView *view, gpointer data);

static gboolean button_press_event_cb (GtkWidget *widget, GdkEventButton *event, gpointer data);

static void paste_done_cb (GtkTextBuffer *buffer, GtkClipboard *clipboard, gpointer data);

static void tab_removed_cb (GeditWindow *window, GeditTab *tab, gpointer data);

static void tab_added_cb (GeditWindow *window, GeditTab *tab, gpointer data);


GEDIT_PLUGIN_REGISTER_TYPE (ScrollpastePlugin, scrollpaste_plugin)


static void
scrollpaste_plugin_init (ScrollpastePlugin *plugin)
{
	plugin->priv = SCROLLPASTE_PLUGIN_GET_PRIVATE (plugin);

	gedit_debug_message (DEBUG_PLUGINS, "ScrollpastePlugin initializing");
}


static void
scrollpaste_plugin_finalize (GObject *object)
{
	gedit_debug_message (DEBUG_PLUGINS, "ScrollpastePlugin finalizing");

	G_OBJECT_CLASS (scrollpaste_plugin_parent_class)->finalize (object);
}


static void
paste_data_free (gpointer data)
{
	PasteData *paste_data = (PasteData *) data;

	if (paste_data->view)
	{
		g_object_unref (paste_data->view);
		paste_data->view = NULL;
	}
	g_free (paste_data);
}


static void
disconnect_from_view (GeditView *view, gpointer data)
{
	GtkTextBuffer *buffer;
	PasteData *paste_data;

	buffer = gtk_text_view_get_buffer (GTK_TEXT_VIEW (view));

	paste_data = g_object_get_data (G_OBJECT (view), WINDOW_DATA_KEY);
	g_signal_handlers_disconnect_by_func (view, button_press_event_cb, paste_data);
	g_signal_handlers_disconnect_by_func (buffer, paste_done_cb, paste_data);

	g_object_set_data (G_OBJECT (view), WINDOW_DATA_KEY, NULL);
}


static void
connect_to_view (GeditView *view, gpointer data)
{
	GtkTextBuffer *buffer;
	PasteData *paste_data;

	paste_data = g_new0 (PasteData, 1);
	paste_data->view = g_object_ref (view);

	g_object_set_data_full (G_OBJECT (view), WINDOW_DATA_KEY, paste_data, paste_data_free);

	g_signal_connect (
		view, "button-press-event", G_CALLBACK (button_press_event_cb), paste_data);

	buffer = gtk_text_view_get_buffer (GTK_TEXT_VIEW (view));
	g_signal_connect (
		buffer, "paste-done", G_CALLBACK (paste_done_cb), paste_data);
}


static gboolean
button_press_event_cb (GtkWidget *widget, GdkEventButton *event, gpointer data)
{
	PasteData *paste_data;
	GtkTextIter iter;
	GtkTextBuffer *buffer;
	gint x, y;

	if (event->button != 2) return FALSE;

	/*
	 * Fix GtkTextView's annoying paste behaviour when pasting with the mouse
	 * (middle button click). By default gtk will scroll the text view to the
	 * original place where the cursor is.
	 */
	paste_data = (PasteData *) data;

	gtk_text_view_window_to_buffer_coords (
		GTK_TEXT_VIEW (widget), GTK_TEXT_WINDOW_TEXT,
		event->x, event->y,
		&x, &y);

	gtk_text_view_get_iter_at_position (
		GTK_TEXT_VIEW (widget), &iter, NULL, x, y);


	buffer = gtk_text_view_get_buffer (GTK_TEXT_VIEW (widget));
	paste_data->paste_mark = gtk_text_buffer_create_mark (
		buffer, NULL, &iter, FALSE);

	return FALSE;
}


static void
paste_done_cb (GtkTextBuffer *buffer, GtkClipboard *clipboard, gpointer data)
{
	PasteData *paste_data;
	GtkTextIter iter;

	paste_data = (PasteData *) data;
	if (paste_data->paste_mark == NULL)
	{
		/* If there's no marker set then it means that the paste wasn't done with
		 * the mouse. That's ok since pasting the normal paste operation works and
		 * doesn't require a hack.
		 */
		return;
	}

	/*
	 * If a paste is done through the middle click then place the cursor at the
	 * end of the pasted text.
	 */
	gtk_text_buffer_get_iter_at_mark (buffer, &iter, paste_data->paste_mark);
	gtk_text_buffer_place_cursor (buffer, &iter);
gtk_text_view_set_cursor_visible (paste_data->view, !gtk_text_view_get_cursor_visible (paste_data->view));
//gtk_text_view_set_right_margin (paste_data->view, 10 + gtk_text_view_get_right_margin (paste_data->view));
	gtk_text_view_scroll_to_mark (paste_data->view, paste_data->paste_mark,
		0.0, FALSE, 0.0, 0.5);
	gtk_text_buffer_delete_mark (buffer, paste_data->paste_mark);
	paste_data->paste_mark = NULL;
}


static void
tab_removed_cb (GeditWindow *window, GeditTab *tab, gpointer data)
{
	disconnect_from_view (gedit_tab_get_view (tab), NULL);
}


static void
tab_added_cb (GeditWindow *window, GeditTab *tab, gpointer data)
{
	connect_to_view (gedit_tab_get_view (tab), NULL);
}


static void
impl_activate (GeditPlugin *plugin, GeditWindow *window)
{
	GList *list;

	gedit_debug (DEBUG_PLUGINS);
	g_signal_connect (window, "tab-added", G_CALLBACK (tab_added_cb), NULL);
	g_signal_connect (window, "tab-removed", G_CALLBACK (tab_removed_cb), NULL);

	list = gedit_window_get_views (window);
	g_list_foreach (list, (GFunc) connect_to_view, NULL);
	g_list_free (list);
}


static void
impl_deactivate (GeditPlugin *plugin, GeditWindow *window)
{
	GList *list;

	gedit_debug (DEBUG_PLUGINS);
	g_signal_handlers_disconnect_by_func (window, tab_added_cb, NULL);
	g_signal_handlers_disconnect_by_func (window, tab_removed_cb, NULL);

	list = gedit_window_get_views (window);
	g_list_foreach (list, (GFunc) disconnect_from_view, NULL);
	g_list_free (list);
}


static void
impl_update_ui (GeditPlugin *plugin, GeditWindow *window)
{
	gedit_debug (DEBUG_PLUGINS);
}


static void
scrollpaste_plugin_class_init (ScrollpastePluginClass *klass)
{
	GObjectClass *object_class = G_OBJECT_CLASS (klass);
	GeditPluginClass *plugin_class = GEDIT_PLUGIN_CLASS (klass);

	object_class->finalize = scrollpaste_plugin_finalize;

	plugin_class->activate = impl_activate;
	plugin_class->deactivate = impl_deactivate;
	plugin_class->update_ui = impl_update_ui;

	g_type_class_add_private (object_class, sizeof (ScrollpastePluginPrivate));
}

